### ISSUE TITLE
CC replies to original mail starts new tickets

### ISSUE THREAD

**Issue by bekre10** [freescout-helpdesk/freescout#2086]
Hello everyone.
A customer sends a ticket and has multiple people in the CC. If those people in the CC reply, it creates a new ticket each time.
Exception: If we reply first and then the people in the CC reply to our email, then it remains the same ticket.
But if they do reply to the original email from the creator, we end up with multiple new tickets.

It happens often that a customer's employee sends an email to us, e.g. to grant access permissions to some folders on their server and CC's his department manager. Then the manager replies to say it's approved or denied and we end up with 2 tickets instead of one.

Is there any way to address this?

**Comment by freescout-helpdesk** [comment]
Unfortunately not.

**Comment by freescout-helpdesk** [comment]
We've double checked and could not reproduce the issue.

**Comment by freescout-helpdesk** [comment]
> It happens often that a customer's employee sends an email to us

Does customer in this case remove the rest of the email body?

**Comment by bekre10** [comment]
Hmmm I can reproduce it any time.
Customer didn't remove anything.

Customer A
Staff A = Sends out an email to our support address and CC to Staff B
Staff B = Receives the copy of the email send by Staff A to our support and replies to it with modifying anything.
Now it creates a second ticket in our FreeScout.

We have many modules installed. Could it possibly be caused by one of them? Or would you rule this out?

**Comment by freescout-helpdesk** [comment]
What email client does your staff use?

Try to disable all non-official modules.

**Comment by bekre10** [comment]
All modules are purchased from the FreeScout Module Directory. So I assume those are all official ones.

Our clients are using Microsoft Outlook, mostly 2019 and very few with 2016.
Some are also using Outlook online from Office365.
And Endpoint for all is Office365. Using local AD with Azure ADSync (Hybrid).

**Comment by theresabugforsure** [comment]
I'm also able to reproduce this issue 100% of the time.
Scenario 1:
- client opens ticket
- receives auto-reply
- client responds to auto-reply
- new ticket is created

Scenario 2:
- client opens ticket
- receives auto-reply
- agent responds to ticket
- client replies to agent
- new ticket is created

**Comment by theresabugforsure** [comment]
I have these modules active.

Custom Mailbox SignaturesActive
Customization & RebrandingActive
End-User PortalActive
Extended EditorActive
Mobile NotificationsActive
Ticket NumberActive
Widgets White-LabelingActive

**Comment by theresabugforsure** [comment]
More info.

Replying from Thunderbird v102 consistently creates a new ticket. 

Replying from Outlook Web adds the reply to the thread and does not create a new ticket.

**Comment by theresabugforsure** [comment]
Also seems to work fine in the previous version of Thunderbird 91.11.0.

What's used in the email to link a reply to an existing conversation that might be getting mangled by Thunderbird 102?

**Comment by bekre10** [comment]
I only have the issue when someone in the CC replies before we sent the first reply.
It's quite an inconvenience. So we set up an auto reply and modified the mail.php file to include the original message. We did this hoping that the other people in the CC will rather reply to this mail then to the original they received.
Why the script in mail.php works fine and does what it is suppose to do, people of course still reply to the initial email causing the creation of new tickets with each reply.

**Comment by avenjamin** [comment]
I'm seeing the same issue - the original email cc'd our freescout mailbox and new conversations are being created for every reply to the original email rather than being appended to the first conversation like a thread.

We now have 4 or more separate conversations with identical subjects and recipients/senders.
There's a few other headers that might be of use - References or Thread-Index (maybe microsoft online)

I came across this on how best to detect a thread in case it helps.
https://www.jwz.org/doc/threading.html

If you need sample messages I can provide.

**Comment by avenjamin** [comment]
Looking at the `threads` SQL table there is a `message_id` field that is filled from the message header.

For every new message that comes in compare the `In-Reply-To` header to all previous message_ids to detect if it's a reply.

Hopefully this should be enough to catch most replies.
You could compare the subject as well as an additional sanity check, but would have to handle the "Re:" and variants that gets prefixed.

**Comment by avenjamin** [comment]
Just found in the code that this is already being done:

https://github.com/freescout-helpdesk/freescout/blob/4e027c715fdf48da7259a0bf773e5ed9a2b4a5e8/app/Console/Commands/FetchEmails.php#L246-L493

For some reason this doesn't appear to be working as expected though.

**Comment by freescout-helpdesk** [comment]
> I'm seeing the same issue - the original email cc'd our freescout mailbox and new conversations are being created for every reply to the original email rather than being appended to the first conversation like a thread.

Can you provide a body of one of the replies which started a new conversation (using `Show Original` menu button).

**Comment by avenjamin** [comment]
@freescout-helpdesk where is this `Show Original` thread dropdown button?

I have the original emails as .eml files that I can send - I downloaded each one from outlook webmail.
I'd prefer not to put them here publicly. Can I send to support@freescout.net instead?

Cheers
Ben

**Comment by freescout-helpdesk** [comment]
> where is this Show Original thread dropdown button?

![2022-12-23_08-29-05](https://user-images.githubusercontent.com/40499291/209276883-c70e2acc-4a89-4fdd-991c-df5ec0609460.png)

Yes. You can send them to support@freescout.net

**Comment by avenjamin** [comment]
Ah - didn't spot the little down arrow.
I've just emailed through 4 emails that should have formed a thread.

Cheers
Ben

**Comment by freescout-helpdesk** [comment]
> Ah - didn't spot the little down arrow. I've just emailed through 4 emails that should have formed a thread.
> 
> Cheers Ben

We've checked those emails - they don't have FreeScout's Message-IDs in `In-Reply-To` and `References` headers. And there is no special FreeScout's Message-ID marker in the HTML body of those emails. It means the person replying was replying not directly to the email sent by FreeScout or his/her mail server did some magic and changed Message-ID of the email sent by FreeScout.

**Comment by bekre10** [comment]
In our case, the issue occurs when
User A sends an email to our support and CCs User B and User C.
Then user B replies to the email before they have received the first reply from us.
After that user C replies to initial email and so does it go on and we have tons of duplicates.

If FreeScout would record the initial Message ID and then put all subsequent messages which include that message ID to the same thread, it might help, I guess.

**Comment by freescout-helpdesk** [comment]
> User A sends an email to our support and CCs User B and User C.
> Then user B replies to the email before they have received the first reply from us.
> After that user C replies to initial email and so does it go on and we have tons of duplicates.

We've tested it - and it works fine. Replies from senders B and C are connected to the conversation created by sender A. What mail service providers and email clients do senders A, B and C use?

**Comment by bekre10** [comment]
They use Microsoft365 for emails, and so do we.
Freescount uses an API key to connect to our support email account on the M365 tenant.

**Comment by avenjamin** [comment]
@freescout-helpdesk can we have another look into this issue.

If you look at the attached screenshot I have over 6 conversations created for the same email.
And another 2 for another conversation.

We're using Microsoft 365.

We need to look at the logic at how Freescout decides if it's a reply or a new conversation.
There appears to be enough information in the headers to show that it's a reply to a conversation already in Freescout.

The `In-Reply-To` header in conversation \#20206 matches the `Message-ID` header in \#20205.
Likewise the `In-Reply-To` header in conversation \#20205 matches the `Message-ID` header in \#20204.
This continues all the way back to the original message.

\#20206
```
Message-ID:	<6A545242-559F-408D-B774-0C5746F6D6E7@redacted.com>
References:	<SY6PR01MB73920657DDBB51257A55527AB6C89@SY6PR01MB7392.ausprd01.prod.outlook.com>
		<8A8C0044-73B8-4022-84C7-AD7E9E8F7522@redacted.com>
		<SY6PR01MB7392AB9F92F75AB65BA59240B6C99@SY6PR01MB7392.ausprd01.prod.outlook.com>
In-Reply-To:	<SY6PR01MB7392AB9F92F75AB65BA59240B6C99@SY6PR01MB7392.ausprd01.prod.outlook.com>
```

\#20205
```
Message-ID:	<SY6PR01MB7392AB9F92F75AB65BA59240B6C99@SY6PR01MB7392.ausprd01.prod.outlook.com>
References:	<SY6PR01MB73920657DDBB51257A55527AB6C89@SY6PR01MB7392.ausprd01.prod.outlook.com>
		<8A8C0044-73B8-4022-84C7-AD7E9E8F7522@redacted.com>
In-Reply-To:	<8A8C0044-73B8-4022-84C7-AD7E9E8F7522@redacted.com>
```
	
\#20204
```
Message-ID:	<8A8C0044-73B8-4022-84C7-AD7E9E8F7522@redacted.com>
References:	<SY6PR01MB73920657DDBB51257A55527AB6C89@SY6PR01MB7392.ausprd01.prod.outlook.com>
In-Reply-To:	<SY6PR01MB73920657DDBB51257A55527AB6C89@SY6PR01MB7392.ausprd01.prod.outlook.com>
```

<img width="1465" alt="freescout cc issue" src="https://user-images.githubusercontent.com/869534/214454924-54627a0f-3b70-42da-8baf-4a17c1b4c2cb.png">

**Comment by avenjamin** [comment]
Digging into the code I can see where you're getting the previous message ID from either the `In-Reply-To` or `References` header:

https://github.com/freescout-helpdesk/freescout/blob/6ab04ba685b8c6eacb1d0c667416847468e9fe0e/app/Console/Commands/FetchEmails.php#L315-L332

Then further down, the code is trying to look for your special marker from a notification email, but it's thats not found then line 480 shows you just do a search in the threads SQL table for the first message_id that matches the previous message id:

https://github.com/freescout-helpdesk/freescout/blob/6ab04ba685b8c6eacb1d0c667416847468e9fe0e/app/Console/Commands/FetchEmails.php#L480

For some reason this is not working or it's being skipped over.
If I search the `threads` table myself for one of  the `In-Reply-To` message IDs in my previous message I can find a thread that matches.

What's your recommended method to add some debugging messages around this code to see what's going on?

Cheers
Ben

**Comment by avenjamin** [comment]
Think I've found the bug and also the fix 😄 

On line 318:

https://github.com/freescout-helpdesk/freescout/blob/6ab04ba685b8c6eacb1d0c667416847468e9fe0e/app/Console/Commands/FetchEmails.php#L318

The `$in_reply_to` value is still wrapped with `<` & `>` where as the `$message_id` has them stripped off.

When searching for a previous thread on line 480 it can't find one due to the extra `<` `>` characters.

https://github.com/freescout-helpdesk/freescout/blob/6ab04ba685b8c6eacb1d0c667416847468e9fe0e/app/Console/Commands/FetchEmails.php#L480

I modified line 318 to be:

```
$prev_message_id = trim($in_reply_to,'<>');
```

All the tests I've run now find the previous thread and append the message to the existing conversation.
Responding to a freescout ticket notification email also still works.

To test this yourself insert the following code onto line 333:
```
$this->info('['.date('Y-m-d H:i:s').'] Previous Message ID: '.$prev_message_id);
```
1) Compose an email to your freescout mailbox and CC another email address.
2) From the other email account reply to the email - NOT the freescout notification!
3) Notice the fetch-emails.log will show: Previous Message ID: <blahblah>
4) Apply the above fix to line 318 trimming the <>
5) Sent another reply to the original email - NOT the freescout notification.
6) Notice the fetch-emails.log will show: Previous Message ID: blahblah
7) Also notice the reply attach to the original conversation
8) Commit the fix :)

Hopefully this helps resolve this issue completely for everyone.
